File: reference/sdk/src/settlement.ts
import { ethers } from "ethers";
import { SettlementProposal } from "./types";

const SETTLEMENT_COORDINATOR_ABI = [
  "function propose(address fromWallet, address toWallet, uint256 amount, bytes32 intentId, uint256 expiresAt) external returns (bytes32)",
  "function accept(bytes32 proposalId) external",
  "function reject(bytes32 proposalId, string calldata reason) external",
  "function execute(bytes32 proposalId) external payable",
  "function getProposal(bytes32 proposalId) external view returns (address, address, uint256, bytes32, uint256, uint8, string)",
  "function checkExpiry(bytes32 proposalId) external",
];

const STATUS_MAP: SettlementProposal["status"][] = [
  "PENDING",
  "ACCEPTED",
  "REJECTED",
  "EXECUTED",
  "EXPIRED",
];

export class SettlementClient {
  private contract: ethers.Contract;

  constructor(address: string, signer: ethers.Signer) {
    this.contract = new ethers.Contract(address, SETTLEMENT_COORDINATOR_ABI, signer);
  }

  async propose(
    fromWallet: string,
    toWallet: string,
    amount: bigint,
    intentId: string,
    ttlSeconds: number = 3600
  ): Promise<string> {
    const expiresAt = Math.floor(Date.now() / 1000) + ttlSeconds;
    const tx = await this.contract.propose(fromWallet, toWallet, amount, intentId, expiresAt);
    const receipt = await tx.wait();

    // Parse proposalId from event
    const iface = new ethers.Interface([
      "event ProposalCreated(bytes32 indexed proposalId, address indexed from, address indexed to, uint256 amount)",
    ]);
    for (const log of receipt.logs) {
      try {
        const parsed = iface.parseLog(log);
        if (parsed && parsed.name === "ProposalCreated") {
          return parsed.args.proposalId;
        }
      } catch {}
    }
    throw new Error("Could not parse proposalId from receipt");
  }

  async accept(proposalId: string): Promise<void> {
    const tx = await this.contract.accept(proposalId);
    await tx.wait();
  }

  async reject(proposalId: string, reason: string): Promise<void> {
    const tx = await this.contract.reject(proposalId, reason);
    await tx.wait();
  }

  async execute(proposalId: string): Promise<void> {
    const proposal = await this.getProposal(proposalId);
    const tx = await this.contract.execute(proposalId, { value: proposal.amount });
    await tx.wait();
  }

  async getProposal(proposalId: string): Promise<SettlementProposal> {
    const [fromWallet, toWallet, amount, intentId, expiresAt, statusNum, rejectionReason] =
      await this.contract.getProposal(proposalId);
    return {
      proposalId,
      from: fromWallet,
      to: toWallet,
      amount: BigInt(amount),
      intentId,
      expiresAt: Number(expiresAt),
      status: STATUS_MAP[Number(statusNum)] ?? "PENDING",
    };
  }
}

export class MultiAgentSettlement extends SettlementClient {}

File: python-sdk/arc402/settlement.py
"""MultiAgentSettlement — proposes and executes agent-to-agent settlements."""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

from web3 import Web3

from .abis import SettlementCoordinator_ABI
from .exceptions import TransactionFailed
from .types import SettlementProposal

if TYPE_CHECKING:
    from web3.contract import Contract
    from eth_account.signers.local import LocalAccount


class MultiAgentSettlement:
    def __init__(self, w3: Web3, address: str, account: "LocalAccount"):
        self._w3 = w3
        self._account = account
        self._contract: Contract = w3.eth.contract(
            address=Web3.to_checksum_address(address),
            abi=SettlementCoordinator_ABI,
        )

    async def propose(
        self,
        from_wallet: str,
        to_wallet: str,
        amount: int,
        intent_id: bytes,
        ttl_seconds: int = 3600,
    ) -> bytes:
        expires_at = int(time.time()) + ttl_seconds
        tx = self._contract.functions.propose(
            Web3.to_checksum_address(from_wallet),
            Web3.to_checksum_address(to_wallet),
            amount,
            intent_id,
            expires_at,
        ).build_transaction(self._tx_params())
        receipt = await self._send(tx)
        for log in receipt.get("logs", []):
            if len(log.get("topics", [])) > 0:
                return bytes(log["topics"][1])
        raise TransactionFailed("Could not extract proposal ID from receipt")

    async def accept(self, proposal_id: bytes) -> str:
        tx = self._contract.functions.accept(proposal_id).build_transaction(
            self._tx_params()
        )
        receipt = await self._send(tx)
        return receipt["transactionHash"].hex()

    async def reject(self, proposal_id: bytes, reason: str) -> str:
        tx = self._contract.functions.reject(
            proposal_id, reason
        ).build_transaction(self._tx_params())
        receipt = await self._send(tx)
        return receipt["transactionHash"].hex()

    async def execute(self, proposal_id: bytes, amount: int) -> str:
        tx = self._contract.functions.execute(proposal_id).build_transaction(
            {**self._tx_params(), "value": amount}
        )
        receipt = await self._send(tx)
        return receipt["transactionHash"].hex()

    async def get_proposal(self, proposal_id: bytes) -> SettlementProposal:
        from datetime import datetime

        raw = self._contract.functions.getProposal(proposal_id).call()
        return SettlementProposal(
            proposal_id=proposal_id.hex(),
            from_wallet=raw[0],
            to_wallet=raw[1],
            amount=raw[2],
            intent_id=raw[3].hex() if isinstance(raw[3], bytes) else raw[3],
            expires_at=datetime.fromtimestamp(raw[4]),
            status=raw[5],
            rejection_reason=raw[6],
        )

    def _tx_params(self) -> dict:
        return {
            "from": self._account.address,
            "nonce": self._w3.eth.get_transaction_count(self._account.address),
            "gas": 400_000,
            "gasPrice": self._w3.eth.gas_price,
            "chainId": self._w3.eth.chain_id,
        }

    async def _send(self, tx: dict) -> dict:
        signed = self._account.sign_transaction(tx)
        tx_hash = self._w3.eth.send_raw_transaction(signed.raw_transaction)
        return self._w3.eth.wait_for_transaction_receipt(tx_hash)
